File: CarFeatures.js
const count_1s = (num) => {
    let count = 0;
    while (num > 0) {
        num &= num - 1;
        count++;
    }
    return count;
};

const solution = (cars) => {
    cars = cars.map((car) => parseInt(car, 2));
    let left = new Array(cars.length).fill(0);
    let right = new Array(cars.length).fill(0);

    let current_idx = 0;
    let front_idx = 1;
    let similar_cars = 0;
    let bin_1s = 0;

    // pass right comparing feature differences using XOR
    while (current_idx < cars.length) {
        if (front_idx == cars.length) {
            break;
        }
        let xor = cars[current_idx] ^ cars[front_idx];
        bin_1s = count_1s(xor);
        if (bin_1s == 1 || bin_1s == 0) {
            similar_cars++;
        }
        front_idx++;
        if (front_idx == cars.length) {
            right[current_idx] = similar_cars;
            current_idx++;
            front_idx = current_idx + 1;
            similar_cars = 0;
        }
    }

    // Pass left comparing feature differences using XOR
    current_idx = cars.length - 1;
    let prev_idx = cars.length - 2;

    while (current_idx > 0) {
        if (prev_idx == -1) {
            break;
        }
        let xor = cars[current_idx] ^ cars[prev_idx];
        if (count_1s(xor) == 1 || count_1s(xor) == 0) {
            similar_cars++;
        }
        prev_idx--;

        if (prev_idx == -1) {
            left[current_idx] = similar_cars;
            current_idx--;
            prev_idx = current_idx - 1;
            similar_cars = 0;
        }
    }

    let result = right.map((num, idx) => {
        return num + left[idx];
    })

    return result;
};

cars = ["100", "110", "010", "011", "100"];

console.log(solution(cars));

File: CarFeatures.py
""" 
    1. Replace all features with their int equivalent
    2. Right pass of XOR operation of all values to each other:
            count number of 1s in the result
            if number of 1's  == 1:
                similar_count ++     
            right[i] = similar_count 
            i++    
    3. Left pass of XOR operation of all values to each other:
            count number of 1s in the result
            if number of 1's  == 1:
                similar_count ++     
            left[i] = similar_count
            i++
    5. Add the left and right arrays, return this answer

"""
""" 
    Brian Kernighan's algorithm for counting 1's in binary string
"""


def count_1s(n):
    count = 0
    while (n):
        n &= (n - 1)
        count += 1
    return count


def solution(cars):
    if len(cars) == 1:
        return [1]

    if len(cars) < 1:
        return 'No cars'

    cars = [int(car, 2) for car in cars]
    right = [0] * len(cars)
    left = [0] * len(cars)

    current_idx = 0
    next_idx = 1
    similar_cars = 0
    while current_idx < len(cars) - 1:
        xor = cars[current_idx] ^ cars[next_idx]
        ones = count_1s(xor)
        if ones == 1 or ones == 0:
            similar_cars += 1
        next_idx += 1
        if next_idx == len(cars):
            right[current_idx] = similar_cars
            current_idx += 1
            next_idx = current_idx + 1
            similar_cars = 0

    current_idx = len(cars) - 1
    prev_idx = len(cars) - 2
    similar_cars = 0
    while current_idx > 0:
        xor = cars[current_idx] ^ cars[prev_idx]
        ones = count_1s(xor)
        if ones == 1 or ones == 0:
            similar_cars += 1
        prev_idx -= 1
        if prev_idx == -1:
            left[current_idx] = similar_cars
            current_idx -= 1
            prev_idx = current_idx - 1
            similar_cars = 0

    result = [a + b for a, b in zip(left, right)]
    return result


cars = ["0011", "0111", "0111", "0110", "0000"]
ans = solution(cars)
print(ans)